### PR TITLE
fix(primitives): guard normalize_p256_s against out-of-range s values

### PR DESCRIPTION
**Pull request by decofe**
ref: `SIGP-156`

Changes `normalize_p256_s` to return `Result<B256, &'static str>`, rejecting `s == 0` and `s >= P256_ORDER` before the `P256_ORDER - s` subtraction. Without this guard, unsigned subtraction wraps in release mode and can produce an invalid scalar that breaks the low-s postcondition.

Adds edge-case tests for `s == 0`, `s == P256_ORDER`, `s > P256_ORDER`, and `s == U256::MAX`.

Prompted by: rusowsky

### CODE REVIEW

**Comment by decofe** [comment]
## :eye: Cyclops Security Review

[`c2e254d`](https://github.com/tempoxyz/tempo/commit/c2e254d152a4)

🧭 Auditing · mode=`normal` · workers 0/3 done (3 left) · verify pending 0

| Worker | Engine | Progress | Status |
|--------|--------|:--------:|--------|
| `pr-3101-w1` | gemini-3.1-pro-preview | 🔍 [thread-1](file:///cyclops-audit/logs/pr-3101/conversations/pi-1773332743-658.json) · · | Running |
| `pr-3101-w2` | amp/deep | ✅ [thread-1](https://ampcode.com/threads/T-019ce2de-514d-714e-9dfc-087b3e9b010b) 🔍 [thread-2](https://ampcode.com/threads/T-019ce2e0-7e7f-719b-a569-eba33acbdf99) · | Running |
| `pr-3101-w3` | gpt-5.4 | 🔍 [thread-1](file:///cyclops-audit/logs/pr-3101/conversations/pi-1773332747-1218.json) · · | Running |

<details><summary>⚙️ Controls</summary>

- 🚀 Keep only 1 remaining iteration per worker after the current work finishes.
- 👀 Keep only 2 remaining iterations per worker after the current work finishes.
- ❤️ Let only worker 1 continue; other workers skip queued iterations.
- 😄 Let only worker 2 continue; other workers skip queued iterations.
- 🎉 End faster by skipping queued iterations and moving toward consolidation.
- 😕 Stop active workers/verifiers now and start consolidation immediately.

</details>

---
<details><summary>📜 5 events</summary>

🔍 `pr-3101-w1` iter 1/3 [audit-ripple.md]
🔍 `pr-3101-w2` iter 1/3 [audit-focused.md]
🔍 `pr-3101-w3` iter 1/3 [audit-deep-focus.md]
✅ `pr-3101-w2` iter 1 — clear | [Thread](https://ampcode.com/threads/T-019ce2de-514d-714e-9dfc-087b3e9b010b)
🔍 `pr-3101-w2` iter 2/3 [audit-ripple.md]

</details>

File: crates/primitives/src/transaction/tt_signature.rs [hunk 1]
```diff
@@ -33,12 +33,17 @@ pub const P256N_HALF: U256 =
...
 /// If s > n/2, we replace it with n - s.
 ///
-/// This function should be called by all P256 signing code before creating
-/// a signature, as the p256 crate does not guarantee low-s signatures.
-pub fn normalize_p256_s(s_bytes: &[u8]) -> B256 {
+/// Returns `None` if `s` is zero or `s >= P256_ORDER` (out of range for a
+/// valid scalar). This function should be called by all P256 signing code
+/// before creating a signature, as the p256 crate does not guarantee low-s
+/// signatures.
+pub fn normalize_p256_s(s_bytes: &[u8]) -> Option<B256> {
```
Comment (fgimenez) [inline crates/primitives/src/transaction/tt_signature.rs]: this return type is inconsistent with the rest of fallible validation/parsing fns in the module, which use `Result<T, &'static str>`, maybe we could align here too?

File: crates/primitives/src/transaction/tt_signature.rs [hunk 2]
```diff
@@ -33,12 +33,17 @@ pub const P256N_HALF: U256 =
...
     let s = U256::from_be_slice(s_bytes);
+    if s.is_zero() || s >= P256_ORDER {
```
Comment (fgimenez) [inline crates/primitives/src/transaction/tt_signature.rs:41]: `verify_p256_signature_internal` does not have this check, should we add it there too?

File: crates/primitives/src/transaction/tt_signature.rs (hunk 1 above)
Comment (0xKitsune) [reply crates/primitives/src/transaction/tt_signature.rs]: +1 seems like returning Result here is more consistent with other validation paths.

File: crates/revm/src/evm.rs [hunk 3]
```diff
@@ -414,7 +414,7 @@ mod tests {
             Ok(WebAuthnSignature {
                 webauthn_data: Bytes::from(webauthn_data),
                 r: alloy_primitives::B256::from_slice(&sig_bytes[0..32]),
-                s: normalize_p256_s(&sig_bytes[32..64]),
+                s: normalize_p256_s(&sig_bytes[32..64]).expect("p256 crate produces valid s"),
```
Comment (0xKitsune) [inline crates/revm/src/evm.rs]: If we update to return `Result` here instead, we can simply propagate this error rather than relying on expect.

Comment (0xrusowsky) [reply crates/revm/src/evm.rs]: addressed in [`7d50e9a` (this PR)](https://github.com/tempoxyz/tempo/pull/3101/commits/7d50e9a511bc765806127fe011cb992d537af432)

File: crates/primitives/src/transaction/tt_signature.rs (hunk 2 above)
Comment (0xrusowsky) [reply crates/primitives/src/transaction/tt_signature.rs:41]: addressed in [`7d50e9a` (this PR)](https://github.com/tempoxyz/tempo/pull/3101/commits/7d50e9a511bc765806127fe011cb992d537af432)

File: crates/primitives/src/transaction/tt_signature.rs (hunk 1 above)
Comment (0xrusowsky) [reply crates/primitives/src/transaction/tt_signature.rs]: addressed in [`7d50e9a` (this PR)](https://github.com/tempoxyz/tempo/pull/3101/commits/7d50e9a511bc765806127fe011cb992d537af432)

File: crates/primitives/src/transaction/tt_signature.rs [hunk 4]
```diff
@@ -810,8 +815,13 @@ fn verify_p256_signature_internal(
     pub_key_y: &[u8],
     message_hash: &B256,
 ) -> Result<(), &'static str> {
-    // High-s value check: reject signatures where s > n/2 to prevent malleability
+    // Reject out-of-range s values (must be 0 < s < P256_ORDER)
     let s_value = U256::from_be_slice(s);
+    if s_value.is_zero() || s_value >= P256_ORDER {
+        return Err("P256 s value out of range");
+    }
+
```
Comment (klkvr) [inline crates/primitives/src/transaction/tt_signature.rs]: this check is already performed by downstream verifiers, removing it to keep PR scoped to `normalize_p256_s` which is only used in tests

**Review by klkvr — APPROVED** [review]
